### PR TITLE
[test] Add CCoinsViewCache Access/Modify/Write tests

### PR DESCRIPTION
**Pull request by ryanofsky**
Add more comprehensive unit tests for CCoinsViewCache. Right now it is hard to refactor caching code or fix bugs in the caching logic because you have to try to mentally enumerate all the different states the cache might be in to make sure a change doesn't cause unintended consequences. The new tests explicitly enumerate relevant cache states, documenting and verifying the behavior in each state, so it will be safer and easier to make changes to the caching code in the future.

### CODE REVIEW

**Comment by laanwj** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/9308/commits/03ffa4cfff29d8273859edb030bc0eb823b26e7b, thanks for adding tests for this important system

**Comment by MarcoFalke** [comment]
Concept ACK 03ffa4cfff29d8273859edb030bc0eb823b26e7b

File: src/test/coins_tests.cpp [hunk 1]
```diff
@@ -415,4 +417,338 @@ BOOST_AUTO_TEST_CASE(ccoins_serialization)
...
+    CheckModifyCoins(VALUE1, VALUE2, PRUNED, ABSENT, DIRTY|FRESH, NO_ENTRY   );
+    CheckModifyCoins(VALUE1, VALUE2, VALUE3, VALUE3, 0          , DIRTY      );
+    CheckModifyCoins(VALUE1, VALUE2, VALUE3, VALUE3, FRESH      , DIRTY|FRESH);
+    CheckModifyCoins(VALUE1, VALUE2, VALUE3, VALUE3, DIRTY      , DIRTY      );
+    CheckModifyCoins(VALUE1, VALUE2, VALUE3, VALUE3, DIRTY|FRESH, DIRTY|FRESH);
+}
+
+void CheckModifyNewCoins(CAmount cache_value, CAmount modify_value, CAmount expected_value, char cache_flags, char expected_flags, bool coinbase)
+{
+    SingleEntryCacheTest test(ABSENT, cache_value, cache_flags);
```
Comment (morcos) [inline src/test/coins_tests.cpp]: Maybe loop over possibilities for base values

File: src/test/coins_tests.cpp [hunk 2]
```diff
@@ -415,4 +417,338 @@ BOOST_AUTO_TEST_CASE(ccoins_serialization)
...
+    test.cache.SelfTest();
+
+    CAmount result_value;
+    char result_flags;
+    GetCoinsMapEntry(test.cache.map(), result_value, result_flags);
+    BOOST_CHECK_EQUAL(result_value, expected_value);
+    BOOST_CHECK_EQUAL(result_flags, expected_flags);
+}
+
+BOOST_AUTO_TEST_CASE(ccoins_write)
```
Comment (morcos) [inline src/test/coins_tests.cpp]: Maybe try a loop of these tests with Child Flags not DIRTY as well

**Comment by morcos** [comment]
ACK (left a couple of comments to slightly increase test coverage)

File: src/test/coins_tests.cpp (hunk 2 above)
Comment (ryanofsky) [inline src/test/coins_tests.cpp]: Added.

**Comment by paveljanik** [comment]
@ryanofsky What about this warning?
```
test/coins_tests.cpp:432:19: warning: unused variable 'DIRTY_FLAGS' [-Wunused-const-variable]
```

Do you plan to use it or can it be removed?

**Comment by ryanofsky** [comment]
I do have some debug code using it, which I might make a PR from later, but its better to remove for now since it's causing this warning. Created to #9435 to remove.

**Pull request merged**